### PR TITLE
Fix #370

### LINKED ISSUE #370: `A8-4-8`: Allow output parameters in assign operators
### Affected rules
 - `A8-4-8`

### Description

Output parameters are required for user defined assign operators, and so should not be reported.

### Example
```cpp
constexpr byte& operator|=(byte& lhs, const byte rhs) noexcept {
  lhs = (lhs | rhs);
  return lhs;
}
```

### PR DESCRIPTION
**Pull request by rvermeulen**
## Description

Resolves #370 

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - A8-4-8

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/common/src/codingstandards/cpp/Operator.qll [hunk 1]
```diff
@@ -264,3 +267,51 @@ class UserOverloadedOperator extends Function {
     not this.isCompilerGenerated()
   }
 }
+
+/**
+ * A `std::basic_istream` class, or something that can be used
+ * as one. Based on the BasicOStreamClass.
+ */
+private class BasicIStreamClass extends Type {
```
Comment (knewbury01) [inline cpp/common/src/codingstandards/cpp/Operator.qll:275]: should this go in FileStreams.qll? and maybe extend the [pre-existing modelling](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/standardlibrary/FileStreams.qll#L26)? (didnt look too closely at what the difference is, as in why the hierarchy on that is narrowed to class template instantiations only on the pre-existing model for what I think is maybe meant to represent the same thing - confirm on that!)

File: cpp/common/src/codingstandards/cpp/Operator.qll [hunk 2]
```diff
@@ -264,3 +267,51 @@ class UserOverloadedOperator extends Function {
...
+    this.getUnspecifiedType() instanceof BasicIStreamClass
+    or
+    this.(Class).getABaseClass() instanceof BasicIStreamClass
+    or
+    this.(ReferenceType).getBaseType() instanceof BasicIStreamClass
+  }
+}
+
+/** An implementation of a stream insertion operator. */
+class StreamInsertionOperator extends Function {
```
Comment (knewbury01) [inline cpp/common/src/codingstandards/cpp/Operator.qll:288]: same here, check if this should go in filestreams and if that can be simplified afterwards , to have the calls use the operators [here](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/standardlibrary/FileStreams.qll#L55)

File: cpp/autosar/test/rules/A8-4-8/test.cpp [hunk 3]
```diff
@@ -37,3 +37,38 @@ void f7(A &a) { // NON_COMPLIANT
...
+
+constexpr A &operator|=(
+    A &lhs,
+    const A &rhs) noexcept { // COMPLIANT - non-member user defined assignment
+                             // operators are considered an exception.
+  return lhs;
+}
+
+enum class byte : unsigned char {};
+constexpr byte &operator|(const byte &lhs, const byte &rhs) noexcept {
```
Comment (knewbury01) [inline cpp/autosar/test/rules/A8-4-8/test.cpp:49]: label as compliant, 
also I am not immediately sure what this tests other than that the rule correctly defines out parameters , maybe double check if it contributes something unique, and if so the compliant label with a note will help that

**Review by knewbury01 — COMMENTED** [review]
few minor questions/comments, other than that the improvements look really good !

File: cpp/common/src/codingstandards/cpp/standardlibrary/FileStreams.qll [hunk 4]
```diff
@@ -23,15 +24,31 @@ class FileStream extends ClassTemplateInstantiation {
 /**
  * A `basic_istream` like `std::istream`
  */
-class IStream extends ClassTemplateInstantiation {
-  IStream() { this.getTemplate().hasQualifiedName("std", "basic_istream") }
+class IStream extends Type {
```
[the remark this answers is not in the document]
Comment (rvermeulen) [reply cpp/common/src/codingstandards/cpp/standardlibrary/FileStreams.qll:27]: Included two queries identified to be impacted by the new `IStream` type.

[the remark this answers is not in the document]
Comment (knewbury01) [reply cpp/common/src/codingstandards/cpp/standardlibrary/FileStreams.qll:27]: nice!

**Review by knewbury01 — APPROVED** [review]
lgtm!

**Pull request merged**